src/ProjectService.Data/ProjectRepository.cs

```diff
@@ -19,85 +19,147 @@ public class ProjectRepository : IProjectRepository
+
+      if (request.UserId.HasValue)
       {
-        return null;
+        if (request.IncludeUsers)
+        {
+          projects = _provider.Projects
+            .Include(pu => pu.Users)
+            .Where(p => p.Users.Any(u => u.UserId == request.UserId.Value));
+        }
+        else
+        {
+          projects = _provider.ProjectsUsers
+            .Where(pu => pu.UserId == request.UserId)
+            .Include(pu => pu.Project)
+            .Select(pu => pu.Project);
+        }
+      }
+
+      if (request.ProjectsIds != null && request.ProjectsIds.Any())
+      {
+        projects = projects.Where(p => request.ProjectsIds.Contains(p.Id));
+      }
+
+      if (request.DepartmentId.HasValue)
+      {
+        projects = projects.Where(p => p.DepartmentId == request.DepartmentId.Value);
       }
 
-      return dbProject;
+      return projects;
+    }
+
+    #endregion
+
+    public ProjectRepository(
+      IDataProvider provider,
+      IHttpContextAccessor httpContextAccessor)
+    {
+      _provider = provider;
+      _httpContextAccessor = httpContextAccessor;
+    }
+
+    public async Task<DbProject> GetAsync(GetProjectFilter filter)
+    {
+      return await CreateGetPredicate(filter).FirstOrDefaultAsync();
+    }
+
+    public async Task<List<DbProject>> GetAsync(Guid departmentId)
+    {
+      return await _provider.Projects.Where(p => p.DepartmentId == departmentId).ToListAsync();
     }
 
-    public IEnumerable<DbProject> Get(Guid departmentId)
+    public async Task<(List<DbProject>, int totalCount)> GetAsync(IGetProjectsRequest request)
     {
-      return _provider.Projects.Where(p => p.DepartmentId == departmentId);
+      IQueryable<DbProject> projects = CreateGetPredicate(request);
+
+      int totalCount = await projects.CountAsync();
+
+      if (request.SkipCount.HasValue)
+      {
+        projects = projects.Skip(request.SkipCount.Value);
+      }
+
+      if (request.TakeCount.HasValue)
+      {
+        projects = projects.Take(request.TakeCount.Value);
+      }
+
+      if (request.IncludeUsers && !request.UserId.HasValue)
+      {
+        projects = projects.Include(p => p.Users);
+      }
+
+      return (await projects.ToListAsync(), totalCount);
     }
 
-    public Guid? Create(DbProject dbProject)
+    public async Task<Guid?> CreateAsync(DbProject dbProject)
     {
       if (dbProject == null)
       {
         return null;
       }
 
       _provider.Projects.Add(dbProject);
-      _provider.Save();
+      await _provider.SaveAsync();
 
       return dbProject.Id;
     }
 
-    public bool Edit(Guid projectId, JsonPatchDocument<DbProject> request)
+    public async Task<bool> EditAsync(Guid projectId, JsonPatchDocument<DbProject> request)
     {
-      DbProject dbProject = Get(new GetProjectFilter { ProjectId = projectId });
+      DbProject dbProject = await _provider.Projects.FirstOrDefaultAsync(p => p.Id == projectId);
```

Comment:
Could you check if dbProject == default?

Reply:
fixed